design/one-pager-controlled-rollout-of-composition-functions.md

```diff
@@ -0,0 +1,223 @@
+
+To allow for multiple active revisions, we will add a new field to the
+`Function` resource called `activeRevisionLimit`. This setting controls how many
+revisions Crossplane will keep active at any given time. Its name mirrors the
+`revisionHistoryLimit` setting and its value must be no greater than the
+`revisionHistoryLimit`. By default, the `activeRevisionLimit` will be 1,
+maintaining today’s behavior. For all package types other than `Function`, 1
+will be the only valid value for `activeRevisionLimit`, since multiple active
+revisions do not make sense for providers or configurations.
+
+For example, to maintain up to four revisions and up to two active revisions,
+the user would create a function resource like the following:
+
+```yaml
+apiVersion: pkg.crossplane.io/v1
+kind: Function
+metadata:
+  name: function-patch-and-transform
+spec:
+  package: xpkg.crossplane.io/crossplane-contrib/function-patch-and-transform:v0.8.2
+  revisionHistoryLimit: 4
+  activeRevisionLimit: 2
+```
+
+When `revisionActivationPolicy` is `Automatic`, the revisions with the highest
+revision numbers (up to the limit) will be active. If a new revision is created
+and `activeRevisionLimit` revisions are already active, the active one with the
+lowest revision number will be deactivated. When `revisionActivationPolicy` is
+`Manual`, `activeRevisionLimit` is ignored by the package manager and it's left
+to users to activate and deactivate revisions as they wish.
+
+We will update the package manager’s runtime to name services after the
+associated package revision rather than the package, and create a service per
+active revision. This is required to allow multiple active revisions to serve
+traffic. The endpoints used by the composition controller to connect to
+functions are already recorded in the function revision resources, so this is a
+natural change (each active function revision will now have a distinct
+endpoint).
+
+One additional change is necessary in the package manager to enable the
+composition changes described below. The package manager needs to copy labels
+from packages to their revisions, so that users can set a label on a function
+and then use it to select the relevant revision in their composition. Labels on
+compositions already work this way (they get copied to composition revisions),
+so this change will make package revisions more similar to composition
+revisions.
+
+### Composition Changes
+
+To enable compositions to refer to specific function revisions, we will add two
+new optional fields to composition pipeline steps: `functionRevisionRef` and
+`functionRevisionSelector`, mirroring the `compositionRevisionRef` and
+`compositionRevisionSelector` found in XRs. These new fields will allow
+compositions to select a specific function revision by name or by label.
+
+For example, a user wishing to roll out a new version of
+`function-patch-and-transform` to only certain XRs may install the function
+using the following manifest:
```

Comment:
one thing i didn't quite catch from the description of this scenario is how the user interacts with `Function` objects to create new revisions. For a given function, the user will only interact with a single `Function` object, right? i.e. If the user wants multiple revisions, they first create a `Function` object with `.spec.package` that points to version 1, then later on they update that same `Function` object to point to version 2. Along with that update, they would also add this `alpha` label at the same time. Is that right?

The example here was making me second guess that flow because the way its described here it sounds like to create a new revision/version, the user creates another new `Function` object - but I don't think that's right 🤔

Reply:
Your initial understanding is correct - the user updates the `.spec.package` (and any labels they want) in the existing object. I've updated the text in the example to clarify that.

Reply:
Awesome, thank you for making that workflow more clear here!